astropy/utils/iers/tests/test_leap_second.py

```diff
@@ -0,0 +1,402 @@
+        fake_file2 = make_fake_file('27 June 2012', tmpdir)
+        # Ignore warnings about possibly expired built-in file.
+        with catch_warnings(iers.IERSStaleWarning):
+            # Between these and the built-in one, the built-in file is best.
+            ls = iers.LeapSeconds.auto_open([fake_file1, fake_file2,
+                                             iers.IERS_LEAP_SECOND_FILE])
+        assert ls.meta['data_url'] == iers.IERS_LEAP_SECOND_FILE
+
+        # But if we remove the built-in one, the least expired one will be
+        # used.
+        # used and we get a warning that it is stale.
+        with catch_warnings(iers.IERSStaleWarning) as w:
+            ls2 = iers.LeapSeconds.auto_open([fake_file1, fake_file2])
+        assert ls2.meta['data_url'] == fake_file2
+        assert ls2.expires == datetime(2012, 6, 27)
+        assert len(w) == 1
+
+
+@pytest.mark.remote_data
+class TestRemoteURLs:
+    # In these tests, the results may be cached.
+    # This is fine - no need to download again.
+    def test_iers_url(self):
+        ls = iers.LeapSeconds.auto_open([iers.IERS_LEAP_SECOND_URL])
+        assert ls.expires > datetime.now()
+
+    def test_ietf_url(self):
+        ls = iers.LeapSeconds.auto_open([iers.IETF_LEAP_SECOND_URL])
+        assert ls.expires > datetime.now()
+
+
+class TestDefaultAutoOpen:
+    """Test auto_open with different _auto_open_files."""
+    def setup(self):
+        self.good_enough = (datetime.now() +
+                            timedelta(179 - iers.conf.auto_max_age))
+        self._auto_open_files = iers.LeapSeconds._auto_open_files.copy()
+
+    def teardown(self):
+        iers.LeapSeconds._auto_open_files = self._auto_open_files
+
+    def remove_auto_open_files(self, *files):
+        """Remove some files from the auto-opener.
+
+        The default set is restored in teardown.
+        """
+        for f in files:
+            iers.LeapSeconds._auto_open_files.remove(f)
+
+    def test_erfa_found(self):
+        # Set huge maximum age such that whatever ERFA has is OK.
+        # Since it is checked first, it should thus be found.
+        with iers.conf.set_temp('auto_max_age', 100000):
+            ls = iers.LeapSeconds.open()
+        assert ls.meta['data_url'] == 'erfa'
+
+    def test_builtin_found(self):
+        # Set huge maximum age such that built-in file is always OK.
+        # If we remove 'erfa', it should thus be found.
+        self.remove_auto_open_files('erfa')
+        with iers.conf.set_temp('auto_max_age', 100000):
+            ls = iers.LeapSeconds.open()
+        assert ls.meta['data_url'] == iers.IERS_LEAP_SECOND_FILE
+
+    def test_fake_future_file(self, tmpdir):
+        fake_file = make_fake_file('28 June 2345', tmpdir)
+        # Try as system file for auto_open, setting auto_max_age such
+        # that any ERFA or system files are guaranteed to be expired.
+        with iers.conf.set_temp('auto_max_age', -100000), \
```

Comment:
`\` is unnecessary.

Reply:
I hate it too, but cannot use `(..., ...)` with `with` statements. Do you think multiple `with` is better?

Reply:
Hmm, maybe https://docs.python.org/3.5/library/contextlib.html#contextlib.ExitStack